### PR TITLE
Smarter default edit widgets with plugins to pick them (take 2)

### PR DESCRIPTION
**Pull request by pvalsecc**
I've fixed https://github.com/qgis/QGIS/pull/3338 to have QgsEditorWidgetSetup to use the correct export macro and to move it in another file.

Now the widgets factories can give a score on how good they could handle
a widget.

Additionaly, plugins can be added to choose a widget factory in function
of an external information. One of them uses a table in PostgresQL to
allow specification of the widget type and configuration.

I took the opportunity to remove a few deprecated method in relation to
this.

### CODE REVIEW

**Comment by pvalsecc** [comment]
@m-kuhn,
As discussed in #3338, I've changed QgsEditorWidgetSetup.

File: python/core/qgsfield.sip [hunk 1]
```diff
@@ -1,3 +1,35 @@
+/** \ingroup core
+ * Holder for the widget type and its configuration for a field.
+ */
+class QgsEditorWidgetSetup
```
Comment (m-kuhn) [inline python/core/qgsfield.sip]: Can this also be moved to qgseditorwidgetsetup.sip

File: src/app/qgsfieldsproperties.cpp [hunk 2]
```diff
@@ -392,9 +392,9 @@ void QgsFieldsProperties::loadRelations()
       if ( nmrel.fieldPairs().at( 0 ).referencingField() != relation.fieldPairs().at( 0 ).referencingField() )
         nmCombo->addItem( QString( "%1 (%2)" ).arg( nmrel.referencedLayer()->name(), nmrel.fieldPairs().at( 0 ).referencedField() ), nmrel.id() );
 
-      QgsEditorWidgetConfig cfg = mLayer->editFormConfig().widgetConfig( relation.id() );
+      const QgsEditorWidgetSetup setup = QgsEditorWidgetRegistry::instance()->findBest( mLayer, relation.id() );
 
-      QVariant nmrelcfg = cfg.value( "nm-rel" );
+      const QVariant nmrelcfg = setup.config().value( "nm-rel" );
```
Comment (m-kuhn) [inline src/app/qgsfieldsproperties.cpp]: There is not really any advantage in making these local variables const

File: src/core/qgseditorwidgetsetup.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
+/***************************************************************************
+               qgsfield.h - Describes a field in a layer or table
+                     --------------------------------------
+               Date                 : 01-Jan-2004
+               Copyright            : (C) 2004 by Gary E.Sherman
+               email                : sherman at mrcc.com
```
Comment (m-kuhn) [inline src/core/qgseditorwidgetsetup.h]: You can put your own name here

**Comment by pvalsecc** [comment]
OK, 2 out of 3 comments fixed.

For the const one, I just disagree. To me const is a good documentation practice and should be mandatory everywhere. Heck, the language should default to const and should give you a keyword if you want non const. I guess I did too much Erlang...

**Comment by m-kuhn** [comment]
That comment was more a remark, I don't think that should be a blocker.
Fact is, throughout the rest of the codebase, const is a means to document the API. Local variables are modified in whatever way as long as they are not a pointer or reference to shared data.

File: src/gui/editorwidgets/core/qgseditorwidgetregistry.h [hunk 4]
```diff
@@ -199,6 +235,7 @@ class GUI_EXPORT QgsEditorWidgetRegistry : public QObject
 
     QMap<QString, QgsEditorWidgetFactory*> mWidgetFactories;
     QMap<const char*, QPair<int, QString> > mFactoriesByType;
+    QgsEditorWidgetAutoConf autoConf;
```
Comment (m-kuhn) [inline src/gui/editorwidgets/core/qgseditorwidgetregistry.h]: Can this be prefixed with an `m` (`mAutoConf`)?

File: python/core/qgseditformconfig.sip [hunk 5]
```diff
@@ -145,15 +145,6 @@ class QgsEditFormConfig
...
      *
-     * @param fieldIdx  The index of the field
-     *
-     * @return The id for the editor widget or a NULL string if not applicable
-     */
-    QString widgetType( int fieldIdx ) const;
-
-    /**
-     * Get the id for the editor widget used to represent the field at the given index
-     *
```
Comment (m-kuhn) [inline python/core/qgseditformconfig.sip]: I think we should go all or nothing here. Either we allow indexed access for all properties or for none.

So I'd say also replace `removeWidgetConfig(idx)` and others or leave the old methods in place.

**Comment by m-kuhn** [comment]
Sorry to keep comments coming, I just never found one big enough block of time to review it all at once.

**Comment by pvalsecc** [comment]
@m-kuhn, no problem, I can understand that. I'm working on interruption as well, those days.

**Comment by pvalsecc** [comment]
@m-kuhn, OK, all your comments are addressed.

**Comment by jef-n** [comment]
See also [QGIS #15496](https://hub.qgis.org/issues/15496)

**Comment by pvalsecc** [comment]
I'm looking at https://hub.qgis.org/issues/15496, @jef-n, but I don't see see why you want me to.

**Comment by jef-n** [comment]
oops, wrong ticket (edited above)

**Comment by pvalsecc** [comment]
@jef-n, this PR solves all the cases of https://hub.qgis.org/issues/15516 but the enum one. This could be added later without too much problem. But I've spent way to much time on this one already to add this now.

**Comment by sbrunner** [comment]
@m-kuhn can we merge this pull request?

**Comment by m-kuhn** [comment]
Merged (again) in 7169079f915f168afc88aa42ccd451be6c3e06a9 in combination with fixup e1f21b003c700f9e317edb66ec0d9de8fecb80d5

**Comment by pvalsecc** [comment]
Thanks @m-kuhn for the fixups and the merge.

**Comment by nyalldawson** [comment]
@pvalsecc

Great work and a fantastic feature! Thanks for being so receptive to adapting to feedback too!

**Comment by nyalldawson** [comment]
@pvalsecc I'm seeing a regression after this merge - now numeric fields default to range widgets, which have a preset min/max range. This is causing numeric values to be truncated to the default maximum of 99. Not sure how to address this - There doesn't seem to be a way to disable the min/max for a spin box, and worse they are restricted to integer values (not longs). Any ideas?

File: src/gui/editorwidgets/qgsdatetimeeditfactory.cpp [hunk 6]
```diff
@@ -114,3 +114,18 @@ QMap<const char*, int> QgsDateTimeEditFactory::supportedWidgetTypes()
   map.insert( QgsDateTimeEdit::staticMetaObject.className(), 10 );
   return map;
 }
+
+unsigned int QgsDateTimeEditFactory::fieldScore( const QgsVectorLayer* vl, int fieldIdx ) const
+{
+  const QgsField field = vl->fields().field( fieldIdx );
+  const QVariant::Type type = field.type();
+  const QgsEditorWidgetConfig config = vl->editFormConfig().widgetConfig( field.name() );
+  if ( type == QVariant::DateTime || config.contains( "field_format" ) )
```
Comment (nyalldawson) [inline src/gui/editorwidgets/qgsdatetimeeditfactory.cpp]: Shouldn't this also consider Date and Time field types?

Comment (pvalsecc) [inline src/gui/editorwidgets/qgsdatetimeeditfactory.cpp]: I don't have a feeling that QgsDateTimeEditWrapper is wired to support QDate or QTime. It looks to me it supports only QString and QDateTime.

**Comment by pvalsecc** [comment]
@nyalldawson, I'll create a PR for the fixes you suggested.

**Comment by strk** [comment]
This code was found responsible of an extreme slowdown in loading tables:
https://github.com/qgis/QGIS/issues/26186

Could anyone partecipating in this thread comment as for where the `editor_widget_styles`  was introduced and is expected to be ? I'd like to optimize its querying but not having any specification about where it should be makes it hard to deal with...

**Comment by pvalsecc** [comment]
Hi @strk. #31858 looks good. But I don't understand your question.

**Comment by strk** [comment]
The question is: what is the usage of the table based on ?
QGIS itself does not create that table, nor it documents where such table should be and what fields it should have. I can think of a few optimization based on changing the expected table schema, but we cannot change expectances if users are already using this table.

Example: is it ok to only use such table if it is in the public schema ? can a user opt-out from searching such table ?

Please comment directly in #31858 so I know which commit  you say (I'm adding more)

**Comment by pvalsecc** [comment]
True, I cannot find any doc. Sad of me. This was done for a customer project, but I don't know if it is still used. @fxprunayre, do you this is a problem if we change the schema of this table?

**Comment by fxprunayre** [comment]
> but I don't know if it is still used. @fxprunayre, do you this is a problem if we change the schema of this table?

Digging in old docs we made at that time "Add the capability to easily plug new widgets in attribute table or form view. This is required for Array type support. cf. #3338.  Add support for database Array type available in Postgres and declared in sqlite by adding List suffix on any type eg. StringList is an array of String in sqlite. #3467 "
So I don't know about the implementation details but I don't see any issue altering that table. This work was done in the context of the GMLAS driver of OGR (https://gdal.org/drivers/vector/gmlas.html) and being able to more easily use the driver's output database in QGIS.

**Comment by strk** [comment]
The implementation only _uses_ a table called `qgis_editor_widget_styles` if it is found in database in any schema, and errors out if the schema containing the table is not in the QGIS connection session's `search_path`. There is no code, in QGIS, _creating_ that table, so what we can change is the _expectance_ of the table structure/location, not its creation (there's none).

What we really need is to talk with someone who actually uses this support, to find out if it's still really needed. Dropping it would be easier than dealing with compatibility...

**Comment by elpaso** [comment]
Sorry if it may sound drastic, but I think that in such situations we should just drop this undocumented, untested and half implemented (-> broken) functionality, unless someone is willing to take the lead and complete the implementation with tests and documentation, no community efforts and funds should be wasted on that.

**Comment by strk** [comment]
A test is actually in place:
https://github.com/qgis/QGIS/blob/final-3_4_12/tests/src/python/test_provider_postgres.py#L567
and
https://github.com/qgis/QGIS/blob/final-3_4_12/tests/testdata/provider/testdata_pg.sql#L468

Funny that the test creates the table in a different way than written in the comment above code using it (different primary key), but the test is there.

**Comment by nyalldawson** [comment]
I'm with @elpaso here -- it's unreasonable to spend PSC funds on this. It should be reverted to fix the speed issue, and PSC funds used to fix bugs rather then finish an implementation which was broken from day one.

**Comment by strk** [comment]
Does it take a PSC vote to decide on this ?

**Comment by strk** [comment]
What concerns me the most is if we're really tacking the right suspect here.
Nobody of the original ticket watchers follwed up in #26186 so we dont' even tell if reverting that work would fix the speedup. I'm afraid the move to github lost some watchers, like original submitter (Mickey Dietrich)

**Comment by strk** [comment]
Alright, I made a test: REMOVING the support for custom widgets, the test I pointed to still succeeds !!

I guess at this point I'd really drop that support, doesn't make sense at all to retain it without even a test

**Pull request closed without merging**